### ISSUE TITLE
New Metric: Binlog Lag

### ISSUE THREAD

**Issue by gordan-bobic** [shatteredsilicon/ssm-submodules#221]
Lag is not always caused by `sql_thread` struggling to keep up - it is often the `io_thread` that is failing to keep up with fetching upstream binary logs into the local relay logs for the sql_thread to execute. This is why [st-sideload-relay](https://github.com/shatteredsilicon/silicon-toolkit/blob/master/bin/st-sideload-relay) was developed.

The problem is that it is not obvious at a glance whether the bottleneck is the `sql_thread` (in which case [parallel replication](https://shatteredsilicon.net/mysql-parallel-replication-pitfalls-group-commit-sync-delay/) can help) or the `io_thread` (in which case [st-sideload-relay](https://github.com/shatteredsilicon/silicon-toolkit/blob/master/bin/st-sideload-relay) can help.

What we need is a way of measuring the distance (in bytes) between the current master binlog position - the difference between the fetched binlog position on the slave (`Relay_Master_Log_File`, `Read_Master_Log_Pos` from the output of `SHOW SLAVE STATUS`) and the latest written binary log position (SHOW BINARY LOGS) on the master.

Binlog Lag is the total number of bytes between the Read_Master_Log_Pos from `SHOW SLAVE STATUS` on the slave and the `File_Size` in `SHOW BINARY LOGS` on the master.. If the two are not currently on the same file, the size of the files between them is added to the binlog lag metric.

Because these come from two separate server (`SHOW BINARY LOGS` on the master and `SHOW SLAVE STATUS` on the slave), they cannot be obtain from the exact same point in time. Fetch the size from the slave first, then from the master. A value less than 0 fails the sanity check and indicates a bug (it shouldn't be possible).

This metric should be added to the `MySQL Replication Delay` graph, with the `Right Y` in bytes, on the `MySQL` -> `MySQL Replication` dashboard.

This probably means that mysqld_exporter will have to gain similar connectivity functionality as [st-sideload-relay](https://github.com/shatteredsilicon/silicon-toolkit/blob/master/bin/st-sideload-relay), (to connect to the master using replication user to query the upstream for `SHOW BINARY LOGS`) and the replication user will thus need `REPLICATION CLIENT` privileges in addition to `REPLICATION SLAVE` privileges. Or do we already store enough information in SSM to infer the master server and compare the binlog position saved for the master? It would be good if this could be done purely in Grafana based on existing information instead of making mysqld_exporter get more data from an additional server.

The data fetching part could end up being tricky to implement, so I'm marking this for .5 release rather than .4.

**Comment by gordan-bobic** [comment]
I just realised this may be simpler than I thought. We don't need access to master data to establish what is really important here.

**Core Objective:**
Figure out whether we are lagging because of the I/O thread being unable to fetch binlogs from the master fast enough. Or to put it another way - have some indication if the root cause of our lag is such that we might benefit from `st-sideload-relay`.

**Approach:**
We have:
- `SHOW SLAVE STATUS` -> Read_Master_Log_Pos
- `SHOW SLAVE STATUS` -> Exec_Master_Log_Pos
- `SHOW SLAVE STATUS` -> Master_Log_File
- `SHOW SLAVE STATUS` -> Relay_Log_File
- `SHOW SLAVE STATUS` -> Seconds_Behind_Master

From this we can infer how far behind with executing the relay log we are.
So we can tell how many bytes behind the relay log "tip" we are, and how many seconds behind master that translates to.

From this we can compute the I/O thread lag in bytes/second. But when I/O thread lag is the problem, that would mean, for example 100 bytes behind for 10,000 seconds, 1/100 seconds. This gets unwieldy to display, so I think we should present it in seconds/byte. The higher this is, the worse the I/O thread lag is.

This is nowhere nearly as good as being able to see specifically how many bytes behind we are on the I/O thread - but it would at least give us some metric to see whether we have a problem, without the complication of the exporter having to connect to both the master and the slave to provide that ideal measurement, This we could do based on the data we already have in prometheus, so it would just be an additional graph on the replication stats dashboard.

So:
Dashboard: MySQL -> MySQL Replication
Graph: "I/O Thread Delay"
Same size and shape as "MySQL Replication Delay", and immediately below it.
Showing: MySQL Replication Delay / KB of relay log remaining (Read_Master_Log_Pos - Exec_Master_Log_Pos (accounting for the edge case if we just rotated the binlog/relay log))

Putting this in the v4 queue just for this extra graph on the replication dashboard. Will move it to experimental after this "cheap" option is implemented for further analysis in the future.

Would this work or am I forgetting something?

**Comment by oblitorum** [comment]
Wait, I just checked this [mysql doc](https://dev.mysql.com/doc/refman/5.7/en/show-slave-status.html)

> In essence, this field measures the time difference in seconds between the replica SQL thread and the replica I/O thread. If the network connection between source and replica is fast, the replica I/O thread is very close to the source, so this field is a good approximation of how late the replica SQL thread is compared to the source. If the network is slow, this is not a good approximation; the replica SQL thread may quite often be caught up with the slow-reading replica I/O thread, so Seconds_Behind_Master often shows a value of 0, even if the I/O thread is late compared to the source. In other words, this column is useful only for fast networks.

Doesn't it mean that `Seconds_Behind_Master` is actually `(timestamp in latest event IO_thread has read) - (timestamp in latest event SQL_thread has processed)`, not `(timestamp in latest event of master binlog) - (timestamp in latest event IO_thread has read)`? Or am I misunderstanding something here?

**Comment by gordan-bobic** [comment]
I don't think that is correct - because it doesn't fit observations. On the system on which we first identified the requirement for st-sideload-relay we saw huge lag reported (hundreds of thousands of seconds), even though the slave was fully up to date with its relay logs. If this was purely the gap between the sql thread position and the latest relay log entry, the slave lag wouldn't be huge when the slave was constantly caught up with the relay log.

Binlog does actually carry the timestamp, and this is a part of how time-lagged replication works.
```
mysqlbinlog binlog.021975
[...]
# at 256
#240302 10:59:11 server id 142022  end_log_pos 491 CRC32 0x268301be     Gtid list [11-120216-1638193678]
# at 491
```

**Comment by oblitorum** [comment]
I see, if the observation says so.

Let me confirm that we want to use `Seconds_Behind_Master / (Read_Master_Log_Pos - Exec_Master_Log_Pos)` to represent the I/O thread lag, the higher this is, the worse the I/O thread lag is. Is that right?

> Dashboard: MySQL -> MySQL Replication
> Graph: "I/O Thread Delay"
> Same size and shape as "MySQL Replication Delay", and immediately below it.
> Showing: MySQL Replication Delay / KB of relay log remaining (Read_Master_Log_Pos - Exec_Master_Log_Pos (accounting for the edge case if we just rotated the binlog/relay log))

And have 2 lines on the dashboard, one is the I/O thread lag we talk about, and the other one showing the relay log remaining (Read_Master_Log_Pos - Exec_Master_Log_Pos). Is that right?

**Comment by gordan-bobic** [comment]
Thinking about it, maybe it could just be one line added to the existing lag graph instead of having a separate graph.
Note that Read_Master_Log_Pos - Exec_Master_Log_Pos could be referring to different binlog files, so that would have to somehow be accounted for - and I'm not sure if that can be done cleanly with just what is already in prometheus.

What about an alternative approach:
```
mysql_slave_status_seconds_behind_master / (mysql_slave_status_relay_log_space / 1024^3)
```
1024^3 or 1G, because default and most commonly used binlog (and thus relay log) size is 1G, so if we only have one relay log in progress at any one time, it means the sql thread is keeping up. So high lag but low relay log space implies that the I/O thread is bottlenecked on fetching relay logs.

High => I/O thread lag
Low => SQL thread lag

Thouhgs?

**Comment by oblitorum** [comment]
Yeah, I think that should work in a similar way.

**Comment by gordan-bobic** [comment]
OK, let's do that. And since there can be multiple lag lines (for multiple replication sources), maybe we should have it on a separate graph, because in that case the graph will become illegible.

**Comment by oblitorum** [comment]
OK, got it

**Comment by oblitorum** [comment]
This is how it looks like on my local environment (there are 2 replication channels)

![image](https://github.com/shatteredsilicon/ssm-submodules/assets/61085039/62398817-cd5c-42a8-9ed6-c6e890954913)

**Comment by gordan-bobic** [comment]
Cool, so that's 992 seconds per GB?

**Comment by oblitorum** [comment]
Yeah, I think that's what it represents.

**Comment by gordan-bobic** [comment]
Looks good. Send a pull request, but don't close this ticket. We'll push the ticket back to exp target because it would be nice to at some point implement the proper, more exact relay lag measure in bytes.
Also, that second graph should be called something a little more distinct, e.g.
`MySQL Relay Lag`

**Comment by oblitorum** [comment]
Oh, right, I forgot about that
